snuba_sdk/query.py

```diff
@@ -83,18 +84,56 @@ def set_granularity(self, granularity: int) -> Query:
         return self._replace("granularity", Granularity(granularity))
 
     def validate(self) -> None:
-        # TODO: Contextual validation. E.g. top level functions in select
-        # require aliases
-        self.match.validate()
-        clauses = [self.select, self.groupby, self.where]
-        for exps in clauses:
-            if exps is not None:
-                for exp in exps:
-                    exp.validate()
-
-        self.limit.validate() if self.limit is not None else None
-        self.offset.validate() if self.offset is not None else None
-        self.granularity.validate() if self.granularity is not None else None
+        # TODO: Contextual validations:
+        # - Must have certain conditions (project, timestamp, organization etc.)
+
+        if not self.select or len(self.select) == 0:
+            raise InvalidQuery("query must have at least one column in select")
+
+        # Top level functions in the select clause must have an alias
+        non_aggregates = []
+        has_aggregates = False
+        for exp in self.select:
+            if isinstance(exp, Function) and not exp.alias:
+                raise InvalidQuery(
+                    f"{exp.translate()} must have an alias in the select"
+                )
+
+            if not isinstance(exp, Function) or not exp.is_aggregate():
+                non_aggregates.append(exp)
+            else:
+                has_aggregates = True
+
+        # Non-aggregate expressions must be in the groupby if there is an aggregate
+        if has_aggregates and len(non_aggregates) > 0:
+            if not self.groupby or len(self.groupby) == 0:
+                raise InvalidQuery(
+                    "groupby must be included if there are aggregations in the select"
+                )
+
+            for group_exp in non_aggregates:
+                if group_exp not in self.groupby:
+                    raise InvalidQuery(
+                        f"{group_exp.translate()} missing from the groupby"
+                    )
+
+        # TODO - It's not clear if this is worth doing. Each of these components was validated
+        # when they were created, so do we need to validate again? It is possible for someone
+        # to override a component with something invalid, but if there is only so much
+        # idiot proofing that can be done with Python.
```

Comment:
If your dataclass is frozen it is immutable. If the collections inside the class are not MutableSequence, MutableMapping etc. they are immutable, if you change the content in place you will at least have a type error.
So you should not need to validate structures that are immutable. Like Match.

Now, when to validate the entire query is a different and more interesting problem. I would rather we decide a validation policy that is comprehensive and we are convinced that is needed and stick to that. This would make it much easier to have a good confidence on the validity status of the query when dealing with that.

I think deciding doing a deep validation when we validate the query class is a good idea. Redoing it when formatting and translating seems a good option to me to catch improper usages of the api.
IF we guarantee these two are done, then everything else in between is a plus and we would not have to worry when maintaining the api.